### ISSUE TITLE
Week 3: Production Training & Knowledge Transfer

### ISSUE THREAD

**Issue by andreifurdui** [andreifurdui/neural-stack#8]
## Production Training & Knowledge Transfer

**Goal**: Master production-grade vision transformer training through implementation of advanced optimization techniques, systematic transfer learning comparison, and knowledge distillation. Build training infrastructure reusable across future modules.

**Dataset**: FGVC Aircraft (Fine-Grained Visual Classification of Aircraft)
- 100 aircraft variant classes (e.g., Boeing 737-700 vs 737-800)
- ~10,000 images, extreme fine-grained classification challenge
- Access: `torchvision.datasets.FGVCAircraft(root, split='train', download=True)`
- Challenge: Requires global context reasoning (ViT strength) vs local texture patterns (CNN bias)

**Target Time**: 12 Hours

**Location**: `neural_stack/training/` (new module), `notebooks/03_production_training.ipynb`

---

## Learning Objectives

- [x] Master production-grade ViT training dynamics and stabilization techniques
- [x] Apply transfer learning strategies systematically across multiple fine-tuning regimes  
- [x] Implement knowledge distillation from CNN teachers for data-efficient training
- [x] Analyze and compare training approaches with statistical rigour
- [x] Build reusable training infrastructure for Weeks 5-8

---

## Phase 1: Production Training Infrastructure (3 hours)

**Objective**: Implement stable, efficient training pipeline with modern optimization techniques.

### Requirements

**1. Learning Rate Scheduling**
Implement a scheduler combining:
- Linear warmup for initial N% of training (prevent gradient explosion)
- Cosine annealing decay to minimum learning rate
- Mathematical formulation:
  - Warmup: `lr(t) = lr_base * (t / t_warmup)` for `t < t_warmup`
  - Cosine: `lr(t) = lr_min + (lr_base - lr_min) * 0.5 * (1 + cos(π * (t - t_warmup) / (T - t_warmup)))` for `t ≥ t_warmup`

**2. Training Stabilization**
- Gradient clipping (research appropriate norm threshold)
- Mixed precision training using `torch.cuda.amp`
- Proper initialization verification

**3. Data Pipeline**
- Training augmentation: RandomResizedCrop, RandomHorizontalFlip, ColorJitter, RandAugment
- Validation: CenterCrop, Normalize
- Efficient data loading with multiple workers

**4. Monitoring & Logging**
- Track: loss, accuracy, learning rate, gradient norms, training time
- Save checkpoints at regular intervals
- Early stopping based on validation performance

### Task

Train your Week 2 ViT from scratch on FGVC Aircraft for 100 epochs. Document all hyperparameters chosen and justify decisions with references to literature.

### Success Criteria

- [x] Training loss decreases monotonically without spikes
- [ ] Validation accuracy reaches >70% (with proper training)
- [x] Gradient norms remain stable throughout training
- [x] Learning rate schedule visualized and validated
- [x] Training completes in reasonable time (<2 hours on single GPU)

**Deliverable**: Training curves, hyperparameter documentation, convergence analysis

---

## Phase 2: Transfer Learning Systematic Comparison (4 hours)

**Objective**: Empirically determine optimal fine-tuning strategy through controlled experiments.

### Three Transfer Learning Strategies

Load pre-trained ViT from timm: `timm.create_model('vit_base_patch16_224', pretrained=True, num_classes=100)`

**Strategy A: Feature Extraction**
- Freeze all backbone parameters
- Train only classification head
- Rationale: Preserve learned representations when data is scarce

**Strategy B: Partial Fine-tuning**
- Freeze first L/2 transformer layers
- Fine-tune last L/2 layers + head
- Rationale: Balance between adaptation and preserving low-level features

**Strategy C: Full Fine-tuning**
- Train entire model with reduced learning rate (typically 10-100× lower)
- Rationale: Maximum adaptation to target domain when sufficient data

### Experimental Protocol

**For each strategy:**
1. Train for fixed budget (suggest 30 epochs)
2. Use identical data augmentation and batch size
3. Tune learning rate independently for each (critical!)
4. Run 3 random seeds, report mean ± std
5. Measure: final accuracy, training time, convergence speed (epochs to 80%)

### Analysis Requirements

**Quantitative**:
- Accuracy comparison with 95% confidence intervals
- Statistical significance testing (paired t-test, p < 0.05)
- Computation cost analysis (FLOPs, memory, wall-clock time)
- Learning curve overlay showing convergence differences

**Qualitative**:
- When does each strategy excel? Develop decision framework
- How does data efficiency differ? (accuracy at 10%, 50%, 100% of data)
- What happens with different domain shifts? (ImageNet → Aircraft)

### Success Criteria

- [x] All three strategies properly implemented and trained
- [ ] Results reported with confidence intervals (3+ runs)
- [ ] Statistical significance determined for all comparisons
- [ ] Best strategy achieves >85% accuracy
- [ ] Clear recommendation: "Use strategy X when..."

**Deliverable**: Comparison table, statistical analysis, decision framework

---

## Phase 3: Knowledge Distillation Implementation (3 hours)

**Objective**: Implement teacher-student distillation and analyze why it improves ViT training.

### Theoretical Foundation

**Derive the distillation loss mathematically:**

Given:
- Student logits: `z_s ∈ ℝ^C`
- Teacher logits: `z_t ∈ ℝ^C`
- True labels: `y ∈ {0,1}^C` (one-hot)
- Temperature: `τ > 0`

Distillation loss combines:
1. **Hard loss** (standard cross-entropy): `L_hard = -∑_i y_i log(softmax(z_s)_i)`
2. **Soft loss** (KL divergence with temperature-scaled predictions):
   - `L_soft = τ² * KL(softmax(z_t/τ) || softmax(z_s/τ))`
3. **Total loss**: `L = α * L_hard + (1-α) * L_soft`

**Your task**: Explain mathematically why τ² scaling factor is necessary (hint: consider gradient magnitudes)

### Implementation Requirements

**Teacher Model**:
- Use ResNet-50 pre-trained on ImageNet, fine-tuned on FGVC Aircraft
- Freeze after fine-tuning (teacher provides fixed soft targets)

**Student Model**:
- Your Week 2 ViT (same architecture as Phase 1)
- Train from scratch using distillation

**Hyperparameters to explore**:
- Temperature τ: Try {1, 2, 4, 8, 16}, find optimal
- Alpha α: Balance hard and soft loss, try {0.1, 0.3, 0.5, 0.7}
- Show impact on final accuracy through ablation study

### Experimental Design

**Baselines**:
1. Student trained without distillation (from Phase 1)
2. Student trained with distillation (your implementation)
3. Teacher performance (upper bound reference)

**Analysis**:
- Does distilled student surpass from-scratch baseline? By how much?
- How close does student get to teacher performance?
- What's the effect of temperature on soft target entropy? Measure and plot
- Why does distillation help ViTs specifically? (Hint: inductive bias transfer)

### Success Criteria

- [ ] Distillation loss correctly implemented with temperature scaling
- [ ] Distilled model outperforms baseline by >10%
- [ ] Temperature sweep analysis completed (5+ values)
- [ ] Mathematical derivation of τ² scaling provided
- [ ] Written explanation: "Distillation helps because..."

**Deliverable**: Implementation, temperature ablation results, theoretical analysis

---

## Phase 4: Comprehensive Comparative Analysis (2 hours)

**Objective**: Synthesize findings into rigorous comparative analysis with statistical support.

### Required Comparisons

**1. Training Regimes**
- From scratch (baseline)
- Transfer learning (best strategy from Phase 2)
- Distillation (optimal temperature from Phase 3)

**2. Metrics Table**

Create comprehensive table:

| Approach | Final Accuracy | Training Time | Convergence (epochs to 80%) | Parameters Trained | GPU Memory |
|----------|---------------|---------------|-----------------------------|--------------------|------------|
| From Scratch | X ± σ | ... | ... | ... | ... |
| Transfer (Strategy) | X ± σ | ... | ... | ... | ... |
| Distillation | X ± σ | ... | ... | ... | ... |

**All accuracy values must include 95% confidence intervals (≥3 runs)**

**3. Statistical Testing**
- Null hypothesis: "Transfer learning and distillation produce equivalent results"
- Alternative: "One approach significantly outperforms"
- Use paired t-test, report p-values
- Bonferroni correction for multiple comparisons

**4. Ablation Studies**

Systematically remove components:
- What if no warmup? (expect instability)
- What if no data augmentation? (expect overfitting)
- What if temperature τ=1 in distillation? (less smooth targets)

Quantify impact of each component on final performance.

### Written Analysis Requirements

**Section 1: Training Dynamics** (~300 words)
- Why does warmup stabilize training? Reference Fisher information matrix
- How does cosine decay compare to step decay empirically?
- What gradient norm patterns indicate healthy training?

**Section 2: Transfer Learning** (~400 words)
- When should practitioners freeze vs. fine-tune?
- How does domain shift affect optimal strategy?
- Comparison to published results (DeiT paper, ViT paper)

**Section 3: Knowledge Distillation** (~400 words)
- Why do ViTs benefit more from distillation than CNNs?
- How does temperature control knowledge transfer?
- When is distillation worth the added complexity?

**Section 4: Recommendations** (~200 words)
- Decision tree: Given dataset size X and compute budget Y, use approach Z
- Production deployment considerations
- Future work: What would you try next?

### Success Criteria

- [ ] All comparisons statistically validated (p-values reported)
- [ ] Results contextualized against published benchmarks
- [ ] Ablation studies quantify importance of each component
- [ ] Written analysis demonstrates deep understanding
- [ ] Recommendations actionable for practitioners

**Deliverable**: Comprehensive comparison report (Jupyter notebook or PDF)

---

## Code Organization & Reusability

**Requirement**: Structure code for reuse in Weeks 5-8 (self-supervised learning, interpretability, comparative analysis).

**Suggested Structure** (non-prescriptive):
```
neural_stack/
├── training/
│   ├── schedulers.py    # LR scheduling utilities
│   ├── losses.py        # Distillation loss, etc.
│   ├── trainer.py       # Reusable training loop
│   └── utils.py         # Logging, checkpointing
├── models/              # From Weeks 1-2
└── data/                # Dataset utilities
```

**Evaluation criterion**: Can you reuse your Week 3 training code in Week 5 without major refactoring?

---

## Final Submission Checklist

**Code**:
- [ ] Training infrastructure (schedulers, loops, logging)
- [ ] Three transfer learning strategy implementations
- [ ] Knowledge distillation implementation
- [ ] All experiments reproducible from saved configs

**Analysis**:
- [ ] Comprehensive comparison table with confidence intervals
- [ ] Statistical significance testing results
- [ ] Ablation study quantifying component importance
- [ ] 1500+ word written analysis addressing all required sections

**Artifacts**:
- [ ] Training curves for all approaches
- [ ] Attention visualizations comparing from-scratch vs transfer vs distilled
- [ ] Temperature ablation study plots
- [ ] Hyperparameter configuration files

---

## Common Pitfalls to Avoid

⚠️ **Using different hyperparameters across comparisons** → Invalidates fair comparison
⚠️ **Single run without confidence intervals** → Results not statistically meaningful
⚠️ **Not tuning LR independently per strategy** → Transfer learning will underperform
⚠️ **Temperature τ=1 in distillation** → Reduces to standard training, no soft targets
⚠️ **Comparing raw accuracy without statistical testing** → Can't claim one approach is better

---

## Research Extensions (Optional)

**If you finish early, explore**:
- Self-distillation (ViT teacher → ViT student)
- Multi-crop distillation (DeiT's approach)
- Progressive unfreezing schedule
- Distillation from multiple teachers (ensemble)
- Layer-wise learning rate decay

---

## Self-Assessment Questions

**Before submitting, ensure you can answer**:

1. Derive the gradient of the soft distillation loss with respect to student logits
2. Why does freezing more layers reduce overfitting but may limit performance?
3. How would you design an experiment to test whether ViT learns different features than ResNet on fine-grained tasks?
4. If distillation doesn't help, what could be wrong? List 5 potential issues
5. How do your results compare to DeiT paper's reported numbers? If different, why?

---

**Time Breakdown**
- Phase 1 (Production training): 3h
- Phase 2 (Transfer learning): 4h  
- Phase 3 (Distillation): 3h
- Phase 4 (Analysis): 2h

**Total**: ~12 hours
